src/main/java/com/redhat/rhjmc/containerjfr/net/web/http/api/v2/CertificatePostHandler.java

```diff
@@ -0,0 +1,168 @@
+import io.vertx.ext.web.handler.impl.HttpStatusException;
+
+class CertificatePostHandler extends AbstractAuthenticatedRequestHandler {
+
+    static final String PATH = "certificates";
+
+    private final Environment env;
+    private final FileSystem fs;
+    private final Logger logger;
+
+    private static final String TRUSTSTORE_DIR = "TRUSTSTORE_DIR";
+
+    private Function<File, FileOutputStream> outputStreamFunction;
+    private CertificateValidator certValidator;
+
+    @Inject
+    CertificatePostHandler(
+            AuthManager auth,
+            Environment env,
+            FileSystem fs,
+            Logger logger,
+            @Named("OutputStreamFunction") Function<File, FileOutputStream> outputStreamFunction,
+            CertificateValidator certValidator) {
+        super(auth);
+        this.env = env;
+        this.fs = fs;
+        this.logger = logger;
+        this.outputStreamFunction = outputStreamFunction;
+        this.certValidator = certValidator;
+    }
+
+    @Override
+    public ApiVersion apiVersion() {
+        return ApiVersion.V2;
+    }
+
+    @Override
+    public HttpMethod httpMethod() {
+        return HttpMethod.POST;
+    }
+
+    @Override
+    public String path() {
+        return basePath() + PATH;
+    }
+
+    @Override
+    public boolean isAsync() {
+        return false;
+    }
+
+    @Override
+    public boolean isOrdered() {
+        return true;
+    }
+
+    @Override
+    @SuppressFBWarnings("RCN_REDUNDANT_NULLCHECK_WOULD_HAVE_BEEN_A_NPE")
+    public void handleAuthenticated(RoutingContext ctx) throws Exception {
+        FileUpload cert = null;
+        for (FileUpload fu : ctx.fileUploads()) {
+            if ("cert".equals(fu.name())) {
+                cert = fu;
+                break;
+            }
+        }
+
+        if (cert == null) {
+            throw new HttpStatusException(
+                    400, "A file named \"cert\" was not included in the request");
+        }
+
+        Path certPath = fs.pathOf(cert.uploadedFileName());
+
+        if (!env.hasEnv(TRUSTSTORE_DIR)) {
+            throw new HttpStatusException(500, "Truststore directory not set");
+        }
+
+        String truststoreDir = env.getEnv(TRUSTSTORE_DIR);
+        Path filePath = fs.pathOf(truststoreDir, cert.fileName()).normalize();
+
+        if (fs.exists(filePath)) {
+            throw new HttpStatusException(409, filePath.toString() + " Certificate already exists");
+        }
+
+        try (InputStream fis = fs.newInputStream(certPath);
+                DataInputStream dis = new DataInputStream(fis);
+                FileOutputStream out = outputStreamFunction.apply(filePath.toFile())) {
+
+            byte[] bytes = new byte[dis.available()];
+            dis.readFully(bytes);
+            ByteArrayInputStream bytestream = new ByteArrayInputStream(bytes);
+            Certificate certificate = certValidator.parseCertificate(bytestream);
+            byte[] buf = certificate.getEncoded();
```

Comment:
`buf` seems unnecessary - if `parseCertificate` is just being used to validate that `bytestream` represents a well-formed certificate, can't we do `out.write(bytestream)` and avoid potentially re-encoding the `Certificate` back into a byte array, since we already have the byte array that encoded it to begin with?

Reply:
I don't think so - if there were some additional garbage bytes then I should hope that parsing the buffer as a certificate should fail.